### ISSUE TITLE
Is there an easy way to upgrade python version in pipenv?

### ISSUE THREAD

**Issue by weqopy** [pypa/pipenv#2482]
There was an error in pipenv after I upgraded python by `brew upgrade python` and deleted old version by `brew cleanup`.
To solve the problem, I have to run `pipenv --rm` and reinstall pipenv by `pipenv install` and this will installs all packages installed before.
So is there an easy way to upgrade python version in pipenv?

**Comment by uranusjr** [comment]
I believe `pipenv install --python=/path/to/your/python` automatically rebuilds the environment.

**Comment by oz123** [comment]
While this really creates the environment, it does not also update the Pipenvfile ...

**Comment by andrewm4894** [comment]
I think if you just do
```
pipenv --python 3.6
```
It will update the pipfile if needed AND apply the change. e.g. for me i got the below:

```
andrewm4894@dev:~/auto_arima$ pipenv --python 3.6
Virtualenv already exists!
Removing existing virtualenv...
Creating a virtualenv for this project...
Pipfile: /home/andrewm4894/auto_arima/Pipfile
Using /usr/bin/python3.6m (3.6.6) to create virtualenv...
⠋Running virtualenv with interpreter /usr/bin/python3.6m
Using base prefix '/usr'
New python executable in /home/andrewm4894/.local/share/virtualenvs/auto_arima-7RAnXuT8/bin/python3.6m
Also creating executable in /home/andrewm4894/.local/share/virtualenvs/auto_arima-7RAnXuT8/bin/python
Installing setuptools, pip, wheel...done.
Setting project for auto_arima-7RAnXuT8 to /home/andrewm4894/auto_arima

Virtualenv location: /home/andrewm4894/.local/share/virtualenvs/auto_arima-7RAnXuT8
```

**Comment by mchwalisz** [comment]
Sorry, it doesn't work for me. Let me show what I mean.

I have a pipenv environment using old python 2.7:
```bash
sh-4.4$ pwd
/tmp/piptest
sh-4.4$ ls
sh-4.4$ pipenv install requests --python 2.7
Creating a virtualenv for this project...
Pipfile: /tmp/piptest/Pipfile
Using /usr/bin/python2.7 (2.7.15) to create virtualenv...
⠋Running virtualenv with interpreter /usr/bin/python2.7
New python executable in /home/chwalisz/.local/share/virtualenvs/piptest-8PNPu6qk/bin/python2.7
Also creating executable in /home/chwalisz/.local/share/virtualenvs/piptest-8PNPu6qk/bin/python
Installing setuptools, pip, wheel...done.
Setting project for piptest-8PNPu6qk to /tmp/piptest

Virtualenv location: /home/chwalisz/.local/share/virtualenvs/piptest-8PNPu6qk
Creating a Pipfile for this project...
Installing requests...
Collecting requests
  Using cached https://files.pythonhosted.org/packages/65/47/7e02164a2a3db50ed6d8a6ab1d6d60b69c4c3fdf57a284257925dfc12bda/requests-2.19.1-py2.py3-none-any.whl
Collecting chardet<3.1.0,>=3.0.2 (from requests)
  Using cached https://files.pythonhosted.org/packages/bc/a9/01ffebfb562e4274b6487b4bb1ddec7ca55ec7510b22e4c51f14098443b8/chardet-3.0.4-py2.py3-none-any.whl
Collecting urllib3<1.24,>=1.21.1 (from requests)
  Using cached https://files.pythonhosted.org/packages/bd/c9/6fdd990019071a4a32a5e7cb78a1d92c53851ef4f56f62a3486e6a7d8ffb/urllib3-1.23-py2.py3-none-any.whl
Collecting certifi>=2017.4.17 (from requests)
  Using cached https://files.pythonhosted.org/packages/df/f7/04fee6ac349e915b82171f8e23cee63644d83663b34c539f7a09aed18f9e/certifi-2018.8.24-py2.py3-none-any.whl
Collecting idna<2.8,>=2.5 (from requests)
  Using cached https://files.pythonhosted.org/packages/4b/2a/0276479a4b3caeb8a8c1af2f8e4355746a97fab05a372e4a2c6a6b876165/idna-2.7-py2.py3-none-any.whl
Installing collected packages: chardet, urllib3, certifi, idna, requests
Successfully installed certifi-2018.8.24 chardet-3.0.4 idna-2.7 requests-2.19.1 urllib3-1.23

Adding requests to Pipfile's [packages]...
Pipfile.lock not found, creating...
Locking [dev-packages] dependencies...
Locking [packages] dependencies...
Updated Pipfile.lock (ab273c)!
Installing dependencies from Pipfile.lock (ab273c)...
  🐍   ▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉ 5/5 — 00:00:01
To activate this project's virtualenv, run pipenv shell.
Alternatively, run a command inside the virtualenv with pipenv run.
```

Everything was fine, as it installed requests using 2.7. Now, I want to **make a decision to update**. So according to you I should run:

```
sh-4.4$ pipenv --python 3.7
Virtualenv already exists!
Removing existing virtualenv...
Creating a virtualenv for this project...
Pipfile: /tmp/piptest/Pipfile
Using /usr/bin/python3.7m (3.7.0) to create virtualenv...
⠋Running virtualenv with interpreter /usr/bin/python3.7m
Using base prefix '/usr'
/usr/lib/python3.7/site-packages/virtualenv.py:1041: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
New python executable in /home/chwalisz/.local/share/virtualenvs/piptest-8PNPu6qk/bin/python3.7m
Also creating executable in /home/chwalisz/.local/share/virtualenvs/piptest-8PNPu6qk/bin/python
Installing setuptools, pip, wheel...done.
Setting project for piptest-8PNPu6qk to /tmp/piptest

Virtualenv location: /home/chwalisz/.local/share/virtualenvs/piptest-8PNPu6qk
Warning: Your Pipfile requires python_version 2.7, but you are using 3.7.0 (/home/c/.local/share/v/p/bin/python).
  $ pipenv check will surely fail.
```
Indeed it created new environment with 3.7, but Pipenv still complains. It will keep on complaining, as it still requires 2.7.

```
sh-4.4$ cat Pipfile
[[source]]
url = "https://pypi.org/simple"
verify_ssl = true
name = "pypi"

[packages]
requests = "*"

[dev-packages]

[requires]
python_version = "2.7"
```

Is the only solution to update it manually in Pipfile? Imho this issue should be reopened.

**Comment by mchwalisz** [comment]
In short, the environment gets rebuilt but it doesn't update `Pipfile` and `Pipfile.lock`.

**Comment by uranusjr** [comment]
You need to change it youself, explicitly. Python version upgrade needs to be a very conscious change, and we will not automate it.

**Comment by mchwalisz** [comment]
Then, please help me understand why creating a new environment with different Python version is so easy and allowed. This creates a very confusing situation with `Pipfile` requiring one version and an existing environment with the different version.

**Comment by uranusjr** [comment]
It is allowed because sometimes, that is exactly what people want to do. We don’t think that is a good idea, so we emit a warning for it. (There are even people complaining there even _is_ a warning; people have all sorts of use cases, and we need to cover all that, and take heat from all sides defending their assessment.) The Python version requirement in Pipfile is a hint, not a restriction; it is however still important.

I’m not saying we’re absolutely doing the right thing right now. If you have a better proposal, feel free to submit a PEEP, and I would be happy to review it. Keep in mind, however: this has philosophy implications, and people pulling on every side would be passionate about it.

**Comment by acdha** [comment]
@uranusjr I just hit this because, like most of the Python world, I'm upgrading from 3.6 to 3.7. It feels like a bug that explicitly opting-in with `pipenv update --python 3.7` does not update the Pipfile. Would you be receptive to an enhancement request which does nothing other than making that one command do what the documentation suggests it will do?

**Comment by wamarobert** [comment]
pipenv will fail, what if someone only has virtualenv

**Comment by techalchemy** [comment]
@wamarobert I’m not really sure I understand your question. Can you elaborate?

@acdha I’d say we are definitely receptive (sorry for the delay) —- probably the best bet is to propose some semantics for interacting with this section of the Pipfile from the CLI via a a PEEP (a short markdown document describing what you want to add and how it should behave) submitted to `/peeps` as a PR. The technical details in this case are pretty simple so it’s just a matter of the api design.

**Comment by CristianoYL** [comment]
This worked for me:

1. Update python version in `Pipfile`
2. run command `pipenv update`

Hint: make sure you're not running the command inside the venv. Otherwise it wouldn't work.

**Comment by gotexis** [comment]
`pipenv install --python=C:\Users\[user_name]\AppData\Local\Programs\Python\Python38\python.exe`

on windows, did seem to work.

**Comment by nthexwn** [comment]
I'm seeing the same issue (Pipfile contains wrong version) when creating a new environment:

   alejohnson@alejohnson-ubuntu:~/projects/ve-image-generator$ pipenv --rm
    Removing virtualenv (/home/alejohnson/.local/share/virtualenvs/ve-image-generator-EFZzX4VJ)…
    alejohnson@alejohnson-ubuntu:~/projects/my-project$ **pipenv --python=3.6**
    Creating a virtualenv for this project…
    Pipfile: /home/alejohnson/projects/my-project/Pipfile
    Using /usr/bin/python3.6 (3.6.7) to create virtualenv…
    ⠏ Creating virtual environment...Using base prefix '/usr'
    New python executable in /home/alejohnson/.local/share/virtualenvs/my-project-EFZzX4VJ/bin/python3.6
    Also creating executable in /home/alejohnson/.local/share/virtualenvs/my-project-EFZzX4VJ/bin/python
    Installing setuptools, pip, wheel...
    done.
    Running virtualenv with interpreter /usr/bin/python3.6
    ✔ Successfully created virtual environment!
    Virtualenv location: /home/alejohnson/.local/share/virtualenvs/my-project-EFZzX4VJ
    Warning: Your Pipfile requires python_version **2.7, but you are using 3.6.7** (/home/a/.local/share/v/v/bin/python).
      $ pipenv --rm and rebuilding the virtual environment may resolve the issue.
      $ pipenv check will surely fail.

@mchwalisz stated that "creating a new environment with different Python version is so easy", so presumably most users are only seeing this issue when upgrading an existing environment.

**Comment by hamidshp** [comment]
i use 
pipenv --python 3.7 and it's worked

**Comment by fbenavides69** [comment]
I first manually updated Pipfile to 3.8 then:
pipenv update --python 3.8

**Comment by sumeetsk07** [comment]
I used emacs -nw Pipfile

**Comment by Elijas** [comment]
You can also
`pipenv --rm`
`pipenv install`

**Comment by Gxmadix** [comment]
I tried to install or update above my first environment but none worked. 
At last I removed the environment by : `pipenv --rm` you can ensure that it is delete at `~/.local/share/virtualenvs` for linux users. If not, you can go ahead and remove it by your self. 
then a `pipenv --python 3.8` resolved all my issues.

**Comment by bfleming-ciena** [comment]
The problem does seem to be related to the version of python.  I am on a mac, and I had two versions of python installed with brew, 3.8 and 3.9.

Amplify was looking for a python3.8/site-packages folder, but when I looked inside my pipenv folder, it was referring to python3.9.

So I ran the commands suggested.  Cd to the top of my project.

pipenv --rm

pipenv --python 3.8 shell

I then got an error from pipenv saying my Pipfile is set for 3.9, but I didn't have a Pipfile at the current directory. So I copied a Pipfile there to make pipenv happy. I think the line where it specifies 3.8 fixed it.

```
[[source]]
name = "pypi"
url = "https://pypi.org/simple"
verify_ssl = true

[dev-packages]

[packages]
boto3 = "*"

[requires]
python_version = "3.8"
```

Once I did that, and I told pipenv I wanted to use 3.8, I'm now able to run amplify push as expected, from the top level of my project, no weirdness.
